src/main/java/ch/njol/skript/structures/StructVariables.java

```diff
@@ -66,26 +78,92 @@ public class StructVariables extends Structure {
 		Skript.registerStructure(StructVariables.class, "variables");
 	}
 
-	private final List<NonNullPair<String, Object>> variables = new ArrayList<>();
+	public static class DefaultVariables implements ScriptData {
+
+		private final List<NonNullPair<String, Object>> variables = new ArrayList<>();
+		private final Deque<Map<String, Class<?>[]>> hints = new ArrayDeque<>();
+
+		public DefaultVariables(Collection<NonNullPair<String, Object>> variables) {
+			this.variables.addAll(variables);
+		}
+
+		@SuppressWarnings("unchecked")
+		public void add(String variable, Class<?>... hints) {
+			if (hints == null || hints.length <= 0)
+				return;
+			if (CollectionUtils.containsAll(hints, Object.class)) // Ignore useless type hint
+				return;
+			// Take top of stack, without removing it
+			Map<String, Class<?>[]> head = this.hints.getFirst();
+			head.put(variable, hints);
```

Comment:
```suggestion
			this.hints.getFirst().put(variable, hints);
```
Could this be simplified since the variable is only used once?